docs/decisions/0002-ledger-balance-enforcement.rst

```diff
@@ -0,0 +1,346 @@
+|                3 |               50 |                         |                                         |                         |
++------------------+------------------+-------------------------+-----------------------------------------+-------------------------+
+
+Note that this approach prevents blocking the 3rd request because the lock is Subsidy-/Ledger-specific rather than
+locking the entire Transaction table.  Performance is improved over coarser table-locking, however it uses a MySQL
+feature in an unintended way which may have unintended consequences after implementation and testing.
+
+Approach 3: Distributed Locks Using Redis
+*****************************************
+
+This approach is similar Proxy Row Locking in that the lock corresponds only to a single Subsidy/Ledger, however it is
+implemented using Redis rather than MySQL:
+
+.. code-block:: python
+
+  with atomic_with_redis_lock(f"lock-subsidy-{subsidy_uuid}"):
+          if (quantity < 0) and ((ledger.balance() + quantity) < 0):
+              # throw error
+          # create a transaction with the given quantity
+
+Here's one possible implementation of atomic_with_redis_lock() using `python-redis-lock`_:
+
+.. code-block:: python
+
+  from redis import Redis
+  import redis_lock
+
+  conn = Redis()
+
+  @contextmanager
+  def atomic_with_redis_lock(lock_name):
+      """
+      Grab a redis lock.
+
+      The lock auto-expires after 60 seconds to prevent an app crash from orphaning locks.  auto_renewal=True handles
+      the edge case of the yielded logic actually needing more than 60 seconds to complete by automatically renewing the
+      expiration timer.
+
+      TODO: skip locking during unit testing.
+      """
+      with redis_lock.Lock(conn, lock_name, expire=60, auto_renewal=True):
+          with transaction.atomic(durable=True):
+              yield
+
+Overview
+********
+
+Table Locking Advantages:
+
+* Relatively easy to conceptualize.
+
+Table Locking Disadvantages:
+
+* Potentially poor performance.
+* Not available during unit tests. (Common disadvantage)
+
+Proxy Row Locking Advantages:
+
+* Good performance.
+
+Proxy Row Locking Disadvantages:
+
+* Non-trivial code, hard to understand and potentially debug.
+* Uses MySQL features in unintended ways which could have unintended consequences.
+* Not available during unit tests. (Common disadvantage)
+
+Distributed Locks Using Redis Advantages:
+
+* Good performance, but sensitive to Redis response time.
+* Simple code.  Easy for future engineers to understand.
+
+Distributed Locks Using Redis Disadvantages:
+
+* Sensitive to Redis going down.
+
+  * Across edX codebases, Redis is only used as a Celery and/or Caching backend, two instances where Redis cannot
+    hard-stop the application.  In other words, we have prior-art of using Redis for mission-critical functionality, but
+    nothing that would ever block the app completely.
+
+* Not available during unit tests. (Common disadvantage)
```

Comment:
I might also argue that this is easier to mitigate than a DB-based lock.

Reply:
By mitigate, you mean easier for the code to get around it in a bad way, right?

Reply:
Ah, no, I mean we can more easily come up with an idea to actually use redis via unit tests, so that we could potentially test it in our unit tests.